File: 2024/01/Go/part2.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

// Calculate the similarity score between two lists
func calculateSimilarityScore(leftList, rightList []int) int {
	similarityScore := 0

	for _, left := range leftList {
		count := 0
		for _, right := range rightList {
			if left == right {
				count++
			}
		}
		similarityScore += left * count
	}

	return similarityScore
}

// Parse the input into two integer lists
func parseInput(data string) ([]int, []int) {
	lines := strings.Split(data, "\n")
	var leftList, rightList []int

	for _, line := range lines {
		if line == "" {
			continue
		}

		var left, right int
		_, err := fmt.Sscanf(line, "%d %d", &left, &right)
		if err == nil {
			leftList = append(leftList, left)
			rightList = append(rightList, right)
		}
	}

	return leftList, rightList
}

func main() {
	data, err := ioutil.ReadFile("../input.txt")
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	// Parse the input into two lists
	leftList, rightList := parseInput(string(data))

	// Calculate the similarity score
	result := calculateSimilarityScore(leftList, rightList)

	// Output the result
	fmt.Printf("Total Similarity Score: %d\n", result)
}

File: 2024/01/Node/part2.js
const fs = require('fs');

// Calculate the similarity score
function calculateSimilarityScore(leftList, rightList) {
    let similarityScore = 0;

    // Iterate over each number in the left list
    leftList.forEach(left => {
        let count = 0;
        
        // Count how many times the number appears in the right list
        rightList.forEach(right => {
            if (left === right) {
                count++;
            }
        });

        // Add the product of the number and its count to the similarity score
        similarityScore += left * count;
    });

    return similarityScore;
}

// Function to read the input from the file
function readInputFile(filePath) {
    const data = fs.readFileSync(filePath, 'utf-8');
    const lines = data.split('\n').filter(line => line.trim() !== '');
    const leftList = [];
    const rightList = [];

    lines.forEach(line => {
        const [left, right] = line.split(' ').map(Number);
        leftList.push(left);
        rightList.push(right);
    });

    return { leftList, rightList };
}

function main() {
    const filePath = './input.txt';
    const { leftList, rightList } = readInputFile(filePath);
    const similarityScore = calculateSimilarityScore(leftList, rightList);
    console.log('Similarity Score:', similarityScore);
}

main();
